### ISSUE TITLE
Pylint rule expression-not-assigned no triggered on list comprehension W0106 (B018)

### ISSUE THREAD

**Issue by kaddkaka** [astral-sh/ruff#20132]
### Summary

Pylint W0106 is triggered on each line of this code:

```py
[x + 1 for x in range(10)]
[1 for _ in range(2)]
[x in range(5) if x > 3 else 0 for x in range(10)]
```

But ruffs corresponding rule `B018` is not triggered on either.

### Version

ruff 0.11.13

**Comment by kaddkaka** [comment]
```py
[100]
```

triggers:
- pylint `W0104` "pointless-statement"
- ruff `B018` "Found useless expresion"

Is `B018` an alias of W0104, W0106, or both?

**Comment by ntBre** [comment]
It looks like these are getting caught by the `contains_effect` check in B018, which marks all comprehensions, among other expressions, as possibly containing side effects:

https://github.com/astral-sh/ruff/blob/9930c94c0971487cb3f1ab9bbc7230acac717aec/crates/ruff_linter/src/rules/flake8_bugbear/rules/useless_expression.rs#L98-L99

https://github.com/astral-sh/ruff/blob/9930c94c0971487cb3f1ab9bbc7230acac717aec/crates/ruff_python_ast/src/helpers.rs#L114-L121

so this looks like an intentional decision to err on the side of false negatives to me. There was some discussion of alternatives in https://github.com/astral-sh/ruff/pull/3455, though.
